### ISSUE TITLE
t.a.internet.UDPClient exception when it calls non-existent reactor.connectUDP method

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6468]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)| @wallrj reported|
|-|-|
|Trac ID|trac#6468|
|Type|defect|
|Created|2013-04-24 16:35:02Z|
|Branch|https://github.com/twisted/twisted/tree/deprecate-udpclient-6468-2|

One document says that UDPClient should call reactor.listenUDP
 * https://twistedmatrix.com/documents/current/core/howto/application.html#auto7

Another says it will attempt to call reactor.connectUDP
 * https://twistedmatrix.com/documents/current/api/twisted.application.internet.html

The second one is what actually happens, but it of course doesn't work.

```
Removing stale pidfile /home/richard/projects/Twisted/branches/twistd.pid
2013-04-24 17:17:35+0100 [-] Log opened.
2013-04-24 17:17:35+0100 [-] twistd 13.0.0 (/usr/bin/python 2.7.3) starting up.
2013-04-24 17:17:35+0100 [-] reactor class: twisted.internet.selectreactor.SelectReactor.
2013-04-24 17:17:35+0100 [-] Traceback (most recent call last):
2013-04-24 17:17:35+0100 [-]   File "/home/richard/projects/Twisted/trunk/bin/twistd", line 14, in <module>
2013-04-24 17:17:35+0100 [-]     run()
2013-04-24 17:17:35+0100 [-]   File "/home/richard/projects/Twisted/trunk/twisted/scripts/twistd.py", line 27, in run
2013-04-24 17:17:35+0100 [-]     app.run(runApp, ServerOptions)
2013-04-24 17:17:35+0100 [-]   File "/home/richard/projects/Twisted/trunk/twisted/application/app.py", line 652, in run
2013-04-24 17:17:35+0100 [-]     runApp(config)
2013-04-24 17:17:35+0100 [-]   File "/home/richard/projects/Twisted/trunk/twisted/scripts/twistd.py", line 23, in runApp
2013-04-24 17:17:35+0100 [-]     _SomeApplicationRunner(config).run()
2013-04-24 17:17:35+0100 [-]   File "/home/richard/projects/Twisted/trunk/twisted/application/app.py", line 390, in run
2013-04-24 17:17:35+0100 [-]     self.postApplication()
2013-04-24 17:17:35+0100 [-]   File "/home/richard/projects/Twisted/trunk/twisted/scripts/_twistd_unix.py", line 230, in postApplication
2013-04-24 17:17:35+0100 [-]     self.startApplication(self.application)
2013-04-24 17:17:35+0100 [-]   File "/home/richard/projects/Twisted/trunk/twisted/scripts/_twistd_unix.py", line 349, in startApplication
2013-04-24 17:17:35+0100 [-]     app.startApplication(application, not self.config['no_save'])
2013-04-24 17:17:35+0100 [-]   File "/home/richard/projects/Twisted/trunk/twisted/application/app.py", line 668, in startApplication
2013-04-24 17:17:35+0100 [-]     service.IService(application).startService()
2013-04-24 17:17:35+0100 [-]   File "/home/richard/projects/Twisted/trunk/twisted/application/service.py", line 282, in startService
2013-04-24 17:17:35+0100 [-]     service.startService()
2013-04-24 17:17:35+0100 [-]   File "/home/richard/projects/Twisted/trunk/twisted/application/internet.py", line 166, in startService
2013-04-24 17:17:35+0100 [-]     self._connection = self._getConnection()
2013-04-24 17:17:35+0100 [-]   File "/home/richard/projects/Twisted/trunk/twisted/application/internet.py", line 184, in _getConnection
2013-04-24 17:17:35+0100 [-]     'connect%s' % (self.method,))(*self.args, **self.kwargs)
2013-04-24 17:17:35+0100 [-] AttributeError: 'SelectReactor' object has no attribute 'connectUDP'

```

Here's the short demo tac (which I'd actually written to demonstrate [#6467](https://github.com/twisted/twisted/issues/6467))
 
```
from twisted.application import service, internet
from twisted.internet.protocol import DatagramProtocol
from twisted.python import log



class ConnectedUdpProtocol(DatagramProtocol):
    def startProtocol(self):
        DatagramProtocol.startProtocol(self)

        self.transport.connect('127.0.0.1', 53)

        message = 'FOO'
        log.msg('SENDING: ' + message)
        self.transport.write(message)


    def connectionRefused(self):
        log.err('CONNECTION REFUSED')



application = service.Application("Demo application")
service = internet.UDPClient(0, ConnectedUdpProtocol())
service.setServiceParent(application)

```

<details><summary>Searchable metadata</summary>

```
trac-id__6468 6468
type__defect defect
reporter__rwall rwall
priority__normal normal
milestone__ 
branch__branches_deprecate_udpclient_6468_2 branches/deprecate-udpclient-6468-2
branch_author__rwall rwall
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1366821302000000 1366821302000000
changetime__1371329879000000 1371329879000000
version__None None
owner__rwall rwall

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve commented|
|-|-|

connectUDP was removed way back in Twisted 10, presumably this was forgotten. It has been broken all this time and nobody noticed, so maybe we can remove it. The alternative is to call listenUDP and connect just after it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

Please deprecate it and we'll delete it later.  Applications may be importing it but not using it, or importing it and subclassing it in a way that makes it work.  Deleting it will break them without warning.

The documentation should be fixed to not suggest UDPClient is a useful thing.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_4"></a>|@wallrj commented|
|-|-|

(In [38306]) Branching to 'deprecate-udpclient-6468'

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_5"></a>|@wallrj removed owner<br>@wallrj set status to `new`|
|-|-|

Ready for review in log:branches/deprecate-udpclient-6468
 1. Deprecated t.a.internet.UDPClient
 2. Added deprecation test
 3. Removed references to UDPClient from the API docs and application
    howto.

Build results:
 * http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/deprecate-udpclient-6468

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_6"></a>|@radix set owner to @wallrj|
|-|-|

This branch is apparently conflicting according to the auto-generated diff, but git was able to merge it fine.

I've taken a look and it looks like you've done things the usual way. I can't find any other places that mention UDPClient.

+1

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_7"></a>|@wallrj commented|
|-|-|

(In [38814]) Branching to 'deprecate-udpclient-6468-2'

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_8"></a>|@wallrj set status to `closed`|
|-|-|

(In [38816]) Merge deprecate-udpclient-6468-2: Deprecate t.a.internet.UDPClient

Author: rwall
Reviewer: radix
Fixes: [#6468](https://github.com/twisted/twisted/issues/6468)

Deprecate t.a.internet.UDPClient, which no longer works because it relies on reactor.conectUDP, an API was removed in Twisted 10.
